homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,539 @@
+            cleaning_template, docked_topic, docked_template, fan_speed_topic,
+            fan_speed_template, set_fan_speed_topic, fan_speed_list,
+            send_command_topic
+        ),
+    ])
+
+
+class MqttVacuum(VacuumDevice):
+    """Representation of a MQTT-controlled vacuum."""
+
+    # pylint: disable=no-self-use
+    def __init__(
+            self, name, supported_features, qos, retain, command_topic,
+            payload_turn_on, payload_turn_off, payload_return_to_base,
+            payload_stop, payload_clean_spot, payload_locate,
+            payload_start_pause, battery_level_topic, battery_level_template,
+            charging_topic, charging_template, cleaning_topic,
+            cleaning_template, docked_topic, docked_template, fan_speed_topic,
+            fan_speed_template, set_fan_speed_topic, fan_speed_list,
+            send_command_topic):
+        """Initialize the vacuum."""
+        self._name = name
+        self._supported_features = supported_features
+        self._qos = qos
+        self._retain = retain
+
+        self._command_topic = command_topic
+        self._payload_turn_on = payload_turn_on
+        self._payload_turn_off = payload_turn_off
+        self._payload_return_to_base = payload_return_to_base
+        self._payload_stop = payload_stop
+        self._payload_clean_spot = payload_clean_spot
+        self._payload_locate = payload_locate
+        self._payload_start_pause = payload_start_pause
+
+        self._battery_level_topic = battery_level_topic
+        self._battery_level_template = battery_level_template
+
+        self._charging_topic = charging_topic
+        self._charging_template = charging_template
+
+        self._cleaning_topic = cleaning_topic
+        self._cleaning_template = cleaning_template
+
+        self._docked_topic = docked_topic
+        self._docked_template = docked_template
+
+        self._fan_speed_topic = fan_speed_topic
+        self._fan_speed_template = fan_speed_template
+
+        self._set_fan_speed_topic = set_fan_speed_topic
+        self._fan_speed_list = fan_speed_list
+        self._send_command_topic = send_command_topic
+
+        self._cleaning = False
+        self._charging = False
+        self._docked = False
+        self._status = 'Unknown'
+        self._battery_level = 0
+        self._fan_speed = 'unknown'
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe MQTT events.
+
+        This method is a coroutine.
+        """
+        @callback
+        def message_received(topic, payload, qos):
+            """Handle new MQTT message."""
+            if topic == self._battery_level_topic:
+                battery_level = self._battery_level_template\
+                    .async_render_with_possible_json_value(
+                        payload,
+                        error_value=None)
+                if battery_level is not None:
+                    self._battery_level = int(battery_level)
```

Comment:
A possible improvement (for the future) would be to use a nested dict to lookup the template, function for modifying the value, and attribute name, from the topic. Then all the topic checks could be replaced with one code block.

Also I think you could just have the whole config as an instance attribute and use the string constants to lookup what you need in the config.

Reply:
That's a good idea